internal/pkg/consul/client.go

```diff
@@ -0,0 +1,385 @@
+
+	configurationMap := configuration.ToMap()
+	keyValues := convertInterfaceToConsulPairs("", configurationMap)
+
+	// Put config properties into Consul.
+	for _, keyValue := range keyValues {
+		exists, _ := client.ConfigurationValueExists(keyValue.Key)
+		if !exists || overwrite {
+			if err := client.PutConfigurationValue(keyValue.Key, []byte(keyValue.Value)); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Puts a full configuration struct into the Registry
+func (client *consulClient) PutConfiguration(configuration interface{}, overwrite bool) error {
+	bytes, err := toml.Marshal(configuration)
+	if err != nil {
+		return err
+	}
+
+	tree, err := toml.LoadBytes(bytes)
+	if err != nil {
+		return err
+	}
+
+	err = client.PutConfigurationToml(tree, overwrite)
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+// Gets the full configuration from Consul into the target configuration struct.
+// Passed in struct is only a reference for decoder, empty struct is ok
+// Returns the configuration in the target struct as interface{}, which caller must cast
+func (client *consulClient) GetConfiguration(configStruct interface{}) (interface{}, error) {
+	var err error
+	var configuration interface{}
+
+	exists, err := client.HasConfiguration()
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, fmt.Errorf("the Registry (Consul) doesn't contain configuration for %s. Expected data at %s", client.serviceId, client.configBasePath)
+	}
+
+	// Update configuration data from Consul using decoder
+	updateChannel := make(chan interface{})
+	errorChannel := make(chan error)
+
+	decoder := client.newConsulDecoder()
+	decoder.Consul = client.consulConfig
+	decoder.Target = configStruct
+	decoder.Prefix = client.configBasePath
+	decoder.ErrCh = errorChannel
+	decoder.UpdateCh = updateChannel
+
+	defer decoder.Close()
+	defer close(updateChannel)
+	defer close(errorChannel)
+	go decoder.Run()
+
+	select {
+	case <-time.After(5 * time.Second):
+		err = errors.New("timeout loading config from client")
+	case ex := <-errorChannel:
+		err = errors.New(ex.Error())
+	case raw := <-updateChannel:
+		configuration = raw
+	}
+
+	return configuration, err
+}
+
+// Sets up a Consul watch for the target key and send back updates on the update channel.
+// Passed in struct is only a reference for decoder, empty struct is ok
+// Sends the configuration in the target struct as interface{} on updateChannel, which caller must cast
+func (client *consulClient) WatchForChanges(updateChannel chan<- interface{}, errorChannel chan<- error, configuration interface{}, watchKey string) {
+	// some watch keys may already have the "/", add it for those that don't
+	if !strings.Contains(watchKey, "/") {
+		watchKey = "/" + watchKey
+	}
+
+	decoder := client.newConsulDecoder()
+	decoder.Consul = client.consulConfig
+	decoder.Target = configuration
+	decoder.Prefix = client.configBasePath + watchKey
```

Comment:
Is `watchKey` the equivalent of WritableKey? From edgex-go:
`dec.Prefix = internal.ConfigRegistryStem + internal.CoreDataServiceKey + internal.WritableKey`

Reply:
Yes. See the examples in the README.md where it passes internal.WritableKey for watchKey